### PR TITLE
Largely complete dead code elimination

### PR DESCRIPTION
**Pull request by ltratt**
This PR largely completes dead code elimination in the final commit (https://github.com/ykjit/yk/commit/49374b42bef49df52786d4f2c4303baaf5bc256a). On the way to doing that, I realised that I had to more carefully think about how we handle `Proxy*` instructions (https://github.com/ykjit/yk/commit/b47ddc07e4c2211076d3cd09e5efdde3672b4f0c). With this, I can successfully run yklua and eliminate dead code in its traces.

### CODE REVIEW

**Comment by vext01** [comment]
Superficially looks OK to me.

File: ykrt/src/compile/jitc_yk/jit_ir/mod.rs [hunk 1]
```diff
@@ -1,9 +1,45 @@
...
+//!      memory used down (and also bypasses issues with representing graph structures in Rust, but
+//!      that's slightly accidental).
+//!
+//!   2. The IR is designed to allow most manipulations to be performed without requiring
+//!      reallocation.
+//!
+//! This second property has some surprising consequences:
+//!
+//!   1. We have to introduce `Proxy*` instructions which semi-transparently "forward" an
+//!      instruction's results on. For example if we IR in memory stored along these lines:
```
Comment (ptersilie) [inline ykrt/src/compile/jitc_yk/jit_ir/mod.rs:16]: "results on"? Also "if we HAVE IR in memory".

File: ykrt/src/compile/jitc_yk/jit_ir/mod.rs [hunk 2]
```diff
@@ -1,9 +1,45 @@
...
+//!        ```
+//!
+//!      i.e. we want to pretend that the `%1` doesn't exist and that the reference to `%1` is
+//!      actually just a constant.
+//!
+//!   2. We never store [Operand]s directly: we only ever store [PackedOperand]s. Whenever we want
+//!      to get an [Operand] we *must* use [PackedOperand::unpack]. `unpack` does the necessary
+//!      deproxying.
+//!
+//!   3. In nearly all cases we should view the IR as deproxied: unless you explicitly intend to
```
Comment (ptersilie) [inline ykrt/src/compile/jitc_yk/jit_ir/mod.rs:40]: If we really never want this, why don't we make `Module::inst` the no proxy version and `Module::inst_with_proxies` the other?

File: ykrt/src/compile/jitc_yk/jit_ir/mod.rs (hunk 1 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/jit_ir/mod.rs]: Fixed in 85ef34ee.

File: ykrt/src/compile/jitc_yk/jit_ir/mod.rs (hunk 2 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/jit_ir/mod.rs:40]: I'm pretty sure we can come up with better names but I'd rather get a bit more experience with the API before worrying *too* much about it.

**Comment by ltratt** [comment]
OK to squash?

**Comment by ptersilie** [comment]
Please squash.

**Comment by ltratt** [comment]
Squashed.

**Comment by ltratt** [comment]
Force pushed a rebase.

**Pull request merged**